x-pack/plugins/security_solution/server/endpoint/routes/metadata/handlers.ts

```diff
@@ -167,52 +181,70 @@ export const getMetadataRequestHandler = function (
   const queryStrategy = await metadataRequestContext.endpointAppContextService
     ?.getMetadataService()
-    ?.queryStrategy(
-      metadataRequestContext.requestHandlerContext.core.savedObjects.client,
-      queryStrategyVersion
-    );
-
+    ?.queryStrategy(metadataRequestContext.savedObjectsClient, queryStrategyVersion);
   const query = getESQueryHostMetadataByID(id, queryStrategy!);
-  const hostResult = queryStrategy!.queryResponseToHostResult(
-    await metadataRequestContext.requestHandlerContext.core.elasticsearch.legacy.client.callAsCurrentUser(
-      'search',
-      query
-    )
-  );
+
+  if (
+    !metadataRequestContext.esClient &&
+    !metadataRequestContext.requestHandlerContext?.core.elasticsearch.client
+  ) {
+    return undefined;
+  }
+
+  const es = (metadataRequestContext?.esClient ??
+    metadataRequestContext.requestHandlerContext?.core.elasticsearch
+      .client) as IScopedClusterClient;
+  const response = await es.asCurrentUser.search<HostMetadata>(query);
+
+  const hostResult = queryStrategy!.queryResponseToHostResult(response.body);
+
   const hostMetadata = hostResult.result;
   if (!hostMetadata) {
     return undefined;
   }
 
-  const agent = await findAgent(metadataRequestContext, hostMetadata);
+  /* Skip agent data if we only need meta data, by doing this
+   **  we can avoid users get 404 error from findAgent and
+   ** therefore unable to get endpoint meta data
+   */
+  if (includeAgentData) {
+    const agent = await findAgent(metadataRequestContext, hostMetadata);
 
-  if (agent && !agent.active) {
-    throw Boom.badRequest('the requested endpoint is unenrolled');
-  }
+    if (agent && !agent.active) {
+      throw Boom.badRequest('the requested endpoint is unenrolled');
+    }
 
-  const metadata = await enrichHostMetadata(
-    hostMetadata,
-    metadataRequestContext,
-    hostResult.queryStrategyVersion
-  );
-  return { ...metadata, query_strategy_version: hostResult.queryStrategyVersion };
+    const metadata = await enrichHostMetadata(
+      hostMetadata,
+      metadataRequestContext,
+      hostResult.queryStrategyVersion
+    );
+    return { ...metadata, query_strategy_version: hostResult.queryStrategyVersion };
+  }
+  return { metadata: hostMetadata, query_strategy_version: hostResult.queryStrategyVersion };
 }
 
 async function findAgent(
   metadataRequestContext: MetadataRequestContext,
-  hostMetadata: HostMetadata
+  hostMetadata: HostMetadata,
+  includeAgentData?: boolean
 ): Promise<Agent | undefined> {
   try {
+    if (metadataRequestContext.esClient == null) {
+      throw new Error(`esClient not found`);
+    }
+
     return await metadataRequestContext.endpointAppContextService
       ?.getAgentService()
-      ?.getAgent(
-        metadataRequestContext.requestHandlerContext.core.elasticsearch.client.asCurrentUser,
-        hostMetadata.elastic.agent.id
-      );
+      ?.getAgent(metadataRequestContext.esClient.asCurrentUser, hostMetadata.elastic.agent.id);
```

Comment:
rollback

Reply:
I change my mind, in order to be compatible with the case when  `metadataRequestContext.requestHandlerContext` may not exist, I'd like to keep the logic here and check one of them  (`metadataRequestContext.esClient` and `metadataRequestContext.requestHandlerContext?.core.elasticsearch.client`) should exist.